keps/sig-api-machinery/4153-declarative-validation/README.md

```diff
@@ -0,0 +1,1709 @@
+JSON Schema value validations. Using CEL to validate APIs has been successfully
+demonstrated by CRD Validation Rules, which are on track for GA in 1.28.
+
+CEL rules will typically be placed on struct or field declarations and access
+multiple fields nested below the level of the type or field where the CEL
+expression is declared, e.g.:
+
+```go
+  
+  //+validationRule="!self.widgetType == 'Component' || !['badname2', 'badname2'].exists(notAllowed, self.componentName.contains(notAllowed))",reason=Forbidden
+  type FizzBuzzSpec struct {
+    Type WidgetType `json...`
+    ComponentName string `json...`
+  }
+```
+
+Note that the `+validationRule` IDL tags supports multiple optional fields after the expression string. The supported fields are:
+
+- `message`: A vailidation failure message string.
+- `messageExpression`: A CEL expression that evalutes to a validation failure message string. Takes precedence over message.
+- `reason`: Must be one of [Required, Forbidden, Invalid, RequestEntityTooLarge].
+- `field`: A JSONPath to the sub-field that a validation failure should be attributed to.
+
+We will need to extend our CEL libraries to make it possible to migrate all the
+validation rules that exist in the Kubernetes API today.
+
+- `isFormat() <bool>` and `validateFormat() <list<string>>` will be added to allow formats to be checked in CEL
+  expression and for format violations to be reported using `messageExpression: "self.validateFormat('ipv6')"`
+- IP and CIDR library:
+  - `isIP(string) bool`
+  - `ip(string) IP`
+  - `IP.is4() bool`
+  - `IP.is6() bool`
+  - `IP.isLoopback() bool`
+  - `isCIDR(string) bool`
+  - `cidr(string) CIDR`
+  - `CIDR.overlaps(CIDR) bool`
+  - `CIDR.containsIP(IP) bool`
+- UTF8 Library (for CRD validation, ManagedFields)
+  - IsValid
+  - IsPrint
+  - IsControl
+- Selectors
+  - `Matches(LabelSelector)`
+
+### Formats
+
+A significant portion of all validation rules in the API check that a field
+value conforms to a particular "format". A prominent example is `metadata.name`
+and `metadata.generateName` validation.
+
+We will extend the [available list of
+formats](https://github.com/kubernetes/kube-openapi/blob/7fbd8d59e5b89f2ca43a5dcececbffc0bb186c37/pkg/validation/strfmt/default.go#L128)
+to cover formats heavily used by the Kubernetes API, namely:
+
+| Format                       | Primary validate uses              |
+| ---------------------------- | ---------------------------------- |
+| 'dns1123subdomain'           | metadata name and generateName     |
+| 'dns1123label'               | metadata name and generateName     |
+| 'dns1035label'               | Scoped names and keys              |
+
+We will add all of these to the supported list of formats in kube-openapi.
+We will also document all supported formats on the Kubernetes website.
+
+Other candidate format types are:
+
+- `APIVersion` (GroupVersion)
+- Quantity
+- Qualified name
```

Comment:
Add more detail here?  Pods resources will need this, so I think it's going to be important, but I haven't spent the time to break down exactly what is needed.

Reply:
We should change "qualified name" to something less confusing.  `maybePrefixedName` or `labelKey` or something

Reply:
I've added links to the explicit Go validations being referred to. These are terms used all over k8s validation, so I was planning on simply exposing most of the functions from `apimachinery/pkg/util/validation`

Reply:
I know "qualified name" is used within the codebase, but it's a terrible name (and it's my fault).